docs/configuration.rst
Configuration Handling
======================

.. versionadded:: 1.0.130

Connexion supports a series of configurations that define how the
application should run or behave.

The way configuration in Connexion is designed tries to be fiel to how
Flask configurations work. We levarage the Flask support to
configurations adding the possibility to use configuration file or
environment variables for Connexion specific settings. That way we
enable you to use Connexion in conjunction with other
extensions. Futhermore, the development cycle of Connexion
applications is as close as possible to using Flask direcly.


Configuration Basics
====================

We provide the `connexion.App.config` which is a subclass of a
dictionary and can be modifified just like any dictionary:

.. code-block:: python

    app = connexion.App(__name__)
    app.config['SPECIFICATION_FILE'] = './swagger.yaml'


The configurations are also possible to come from environment
variables. The lookup order of configurations is:

- Set directly in the ``connexion.App#config`` configuration
  attribute;
- Set while calling ``connexion.App`` or ``connexion.App.add_api``
  directly;
- An environment variable with same name of the configuration prefixed
  by ``CONNEXION_``;
- Included in a configuration file passed to ``connexion.App`` or
  ``connexion.App.add_api``;


Builtin Configuration Values
----------------------------

The following configuration values are used internally by Connexion:

.. tabularcolumns:: |p{6.5cm}|p{8.5cm}|

========================================= =========================================
``SPECIFICATION_FILE``                    OpenAPI specification file path. Defaults
                                          to: None
``SPECIFICATION_DIR``                     Directory to use as bases to look for
                                          OpenAPI specification files. Defaults to:
                                          ""
``HOST``                                  The host interface to bind on when server
                                          is running. Defaults to: "0.0.0.0"
``PORT``                                  The port to listen to when server is
                                          running. Defaults to: 5000
``WSGI_SERVER_CONTAINER``                 Which WSGI server container to use.
                                          Options are: "flask", "gevent", and
                                          "tornado". Defaults to: "flask"
``DEBUG``                                 Whether to execute the application in
                                          debug mode. Defaults to: False
``AUTHENTICATE_NOT_FOUND_URLS``           Whether to authenticate paths not defined
                                          in the API specification. Defaults to:
                                          False
``VALIDATE_RESPONSES``                    Whether to validate the operation handler
                                          responses against the API specification
                                          or not. Defaults to: False
``STRICT_PARAMETERS_VALIDATION``          Whether to allow or not additional
                                          parameters in querystring and formData
                                          than the defineds in the API
                                          specification. Defaults to: False
``MAKE_SPEC_AVAILABLE``                   Whether to make available the OpenAPI
                                          sepecification under
                                          CONSOLE_UI_PATH/swagger.json path.
                                          Defaults to: True
``JSON_ENCODER``                          Defines which encoder to use when
                                          serializing objects to JSON. Defaults to:
                                          "connexion.decorators.produces.JSONEncode
                                          r"
``STUB_NOT_IMPLEMENTED_ENDPOINTS``        Returns status code 501, and "Not
                                          Implemented Yet" payload, for the
                                          endpoints which handlers are not found.
                                          Defaults to: False
``MOCK_API``                              Returns example data for all endpoints or
                                          for which handlers are not found. Options
                                          are: None, "all" or "notimplemented".
                                          Defaults to: None
``CONSOLE_UI_AVAILABLE``                  Whether to make the OpenAPI console
                                          available under CONSOLE_UI_PATH config
                                          path. Notice that this overrides the
                                          MAKE_SPEC_AVAILABLE configuration since
                                          the specification is required to be
                                          available to show the console UI.
                                          Defaults to: True
``CONSOLE_UI_PATH``                       Path to mount the OpenAPI console.
                                          Defaults to: "/ui"
========================================= =========================================


Using Environment Variables for Configurations
----------------------------------------------

Connexion automatically look for configurations in the environment
variables of the running system. For that you have to set the
environment variable in the format ``CONNEXION_{{CONFIG_NAME}}``. For
example, to set the port that Connexion should use to run the
application you should set the environment variable
``CONNEXION_PORT`` to 8080.

.. code-block:: bash

    $ export CONNEXION_PORT=8080
    $ connexion run swagger.yaml -v
    INFO:werkzeug: * Running on http://0.0.0.0:8080/ (Press CTRL+C to quit)


Using File for Configurations
-----------------------------

You can generate a base configuration file using the Connexion CLI:

.. code-block:: bash

    $ connexion default-config > myapp.conf


connexion/config.py
from derpconf.config import Config


BASE_API_SECTION = 'Basic API configurations'

Config.define('SPECIFICATION_FILE', None,
              'OpenAPI specification file path.',
              BASE_API_SECTION)

Config.define('SPECIFICATION_DIR', '',
              'Directory to use as bases to look for OpenAPI specification files.',
              BASE_API_SECTION)


SERVER_CONFIG_SECTION = 'Application Server configurations'

Config.define('HOST', '0.0.0.0',
              'The host interface to bind on when server is running.',
              SERVER_CONFIG_SECTION)

Config.define('PORT', 5000,
              'The port to listen to when server is running.',
              SERVER_CONFIG_SECTION)

Config.define('WSGI_SERVER_CONTAINER', 'flask',
              'Which WSGI server container to use. '
              'Options are: "flask", "gevent", and "tornado".',
              SERVER_CONFIG_SECTION)

Config.define('DEBUG', False,
              'Whether to execute the application in debug mode.',
              SERVER_CONFIG_SECTION)


API_CONFIG_SECTION = 'API configurations'

Config.define('AUTHENTICATE_NOT_FOUND_URLS', False,
              'Whether to authenticate paths not defined in the API specification.',
              API_CONFIG_SECTION)

Config.define('VALIDATE_RESPONSES', False,
              'Whether to validate the operation handler responses against '
              'the API specification or not.',
              API_CONFIG_SECTION)

Config.define('STRICT_PARAMETERS_VALIDATION', False,
              'Whether to allow or not additional parameters in querystring '
              'and formData than the defineds in the API specification.',
              API_CONFIG_SECTION)

Config.define('MAKE_SPEC_AVAILABLE', True,
              'Whether to make available the OpenAPI sepecification under '
              'CONSOLE_UI_PATH/swagger.json path.',
              API_CONFIG_SECTION)

Config.define('JSON_ENCODER', 'connexion.decorators.produces.JSONEncoder',
              'Defines which encoder to use when serializing objects to JSON.',
              API_CONFIG_SECTION)

Config.define('STUB_NOT_IMPLEMENTED_ENDPOINTS', False,
              'Returns status code 501, and "Not Implemented Yet" payload, for '
              'the endpoints which handlers are not found.',
              API_CONFIG_SECTION)

Config.define('MOCK_API', None,
              'Returns example data for all endpoints or for which handlers'
              ' are not found. Options are: None, "all" or "notimplemented".',
              API_CONFIG_SECTION)


CONSOLE_UI_SECTION = 'API Console UI configurations'

Config.define('CONSOLE_UI_AVAILABLE', True,
              'Whether to make the OpenAPI console available under '
              'CONSOLE_UI_PATH config path. Notice that this '
              'overrides the MAKE_SPEC_AVAILABLE configuration since '
              'the specification is required to be available to show '
              'the console UI.',
              CONSOLE_UI_SECTION)

Config.define('CONSOLE_UI_PATH', '/ui',
              'Path to mount the OpenAPI console.',
              CONSOLE_UI_SECTION)


def _print_doc_table():
    """ Prints the table to be used in the `docs/configuration.rst`
    documentation file.
    """
    from textwrap import fill
    config = Config()
    column_width = 41
    header = '=' * column_width
    header = '{} {}'.format(header, header)
    newline_spacing = '\n' + (' ' * (column_width + 1))
    print(header)
    for group in config.class_groups:
        keys = config.class_group_items[group]
        for key in keys:
            default_value = config.class_defaults[key]
            if isinstance(default_value, str):
                default_value = '"{}"'.format(default_value)

            desc = '{} Defaults to: {}'.format(
                config.class_descriptions[key], default_value)
            desc = fill(desc, width=column_width).replace('\n', newline_spacing)

            print('``{}``{}{}'.format(key, ' ' * (column_width - len(key) - 3), desc))
    print(header)

if __name__ == '__main__':
    _print_doc_table()
